### ISSUE TITLE
Proposal to change tokenizers defaults.

### ISSUE THREAD

**Issue by Narsil** [huggingface/tokenizers#409]
This proposal is to change `Tokenizers` argument from filenames for the real input that the tokenizer's need.
So this :

```python                                                                                                                          
tokenizer = CharBPETokenizer('tokenizer.json')
```

becomes this
```python                                                                          
tokenizer = CharBPETokenizer.from_file('tokenizer.json')
```
and we add this 
```python
vocab = ["a", "b", "ab"]
merges = [("a", "b")]                                                                          
tokenizer = CharBPETokenizer(vocab, merges)
```

We need this if we have multiple ways of loading (which is the case because we use a single file instead of two now for instance).
Another point, is that for the upcoming 'Unigram` model, we will have 2/3 ways of loading a model.

1) The regular way, from our saved models `SentencePIeceUnigramTokenizer.from_file('mytokenizer.json')`
2) From an already trained spm model for which we want 100% compatiblity `SentencePieceUnigramTokenizer.from_spm('mytokenizer.model')`
3) Custom based, for when we want to use a specific vocab. For instance some tokenizers from spm in the wild, have no correct normalization for nfkc (it seems to be the case for right to left unicode for instance). Currently there is no way to fix it with SPM AFAIK. We could provide a solution by preserving the vocab, and enabling a researcher to change the normalizer efficiently
```python
import sentencepiece as spm
sp = spm.SentencePieceProcessor()
sp.Load('mytokenizer.model')
vocab =  [(sp.id_to_piece(i), sp.get_score(i)) for i in range(sp.piece_size())]

tokenizer = SentencePieceUnigramTokenizer(vocab, sp.unk_id()) 
# Now this tokenizer uses the default [Nmt(), Nkfc()] normalizer instead of the embeded normalizer within `mytokenizer.model`
```

As a side note, we also probably need:
```python                                                                          
tokenizer = CharBPETokenizer.from_pretrained('tokenizer')
```

It's quite a bit of changed, which we should not change after 1.0.

**Comment by n1t0** [comment]
Indeed we definitely need to support providing inputs from in-memory data, as opposed to just files. 
**TLDR;** Changing the constructors on the `Model`s should be enough. (Explaination below)

There are two kinds of tokenizers in the Python bindings at the moment: 
 1. The `Tokenizer` class and all the components that can be attached to it. All of these exist in Rust and have some Python bindings. Their API is already pretty good:
    - We can construct them manually, and call `.save("my-tokenizer.json")` to save them to a single file.
    - We can load one from one of these file by doing `tokenizer = Tokenizer.from_file("my-tokenizer.json")`
    - Or a string (in-memory): `tokenizer = Tokenizer.from_str(tok_str)`
    - Or a buffer (in-memory): `tokenizer = Tokenizer.from_buffer(buffer)`  
 2. The `implementations` that actually provide a way to construct a `Tokenizer` without manually specifying all the parts. These are completely implemented in Python and work as a thin wrapper over the real tokenizer and basically provide a shortcut to create some common tokenizers, or import common ones created with some other libraries.

For the `implementations`, the goal is really to provide a way to either:
 - Train one of these, as they are quite common. In this case we don't really need to load anything, just train and save our brand new tokenizer with `.save("my-tokenizer.json")`.
 - Load one, either from files saved with previous versions of the library, or files that were created using some other library (`subword-nmt`, `fastBPE`, `SentencePiece`, ...). In this case, we need to be able to load with the relevant files for this tokenizer => this is determined by the underlying `Model` (`BPE`, `Unigram` and `WordPiece` mostly). So, generally it will be some kind of vocab or vocab/merges type of file. Now, even in this case, I think the optimal way to keep the tokenizer for later is to call `.save("my-tokenizer.json")`, and treat it as any other tokenizer. This is the best way to ensure that the tokenizer is saved entirely, embeds all the necessary information, and behaves the same with the new upcoming versions of the library.

Really, **in every case**, the canonical way to load a tokenizer should be `tok = Tokenizer.from_file("my-tokenizer.json")`, and we should make it clear in every examples, documentation, README, ...

Anyway, concerning the loading part as described in this issue, we just need to update all the `Model` to be able to load them with both in-memory data, and from files. I think we can even probably keep the exact same API (TBD), and just allow each argument to be either a string (file) or a list/map/other for the in-memory version as such:
```python
# BPE
def __new__(vocab: Union[str, Mapping[str, int]], merges: Union[str, List[str]]):
# WordPiece
def __new__(vocab: Union[str, List[str]]):
[...]
```
This will improve both loading `implementations` and manually constructing instances of `Tokenizer`.

Finally, two notes on the other proposed method to load `implementations`:
 - `from_file("tokenizer.json")` doesn't make sense here. Each of the various `implementations` provide all the components, and just need a `Model`. `Tokenizer.from_file` is the right way to load one of these files as they already contain all the necessary information. (ie; what would happen when doing `SentencePieceUnigramTokenizer.from_file("roberta.tokenizer")`?)
 - The same applies to `from_pretrained`: it only makes sense on `Tokenizer`. Otherwise we can also do things like `BertWordPieceTokenizer.from_pretrained("gpt-2")`.
